common/src/main/java/com/microsoft/identity/common/internal/ui/webview/challengehandlers/SmartcardNfcCertBasedAuthChallengeHandler.java

```diff
@@ -0,0 +1,175 @@
+                    @Override
+                    public void onCreateConnection() {
+                        mDialogHolder.showDialog(new SmartcardNfcLoadingDialog(mActivity));
+                        mSmartcardCertBasedAuthManager.requestDeviceSession(new AbstractSmartcardCertBasedAuthManager.ISessionCallback() {
+                            @Override
+                            public void onGetSession(@NonNull final ISmartcardSession session) throws Exception {
+                                tryUsingSmartcardWithPin(pin, certDetails, request, session);
+                                clearPin(pin);
+                            }
+
+                            @Override
+                            public void onException(@NonNull final Exception e) {
+                                indicateGeneralException(methodTag, e);
+                                request.cancel();
+                                clearPin(pin);
+                                mSmartcardCertBasedAuthManager.stopDiscovery(mActivity);
+                            }
+                        });
+                    }
+
+                    @Override
+                    public void onClosedConnection() {
+                        //Nothing needed.
+                    }
+                });
+                mSmartcardCertBasedAuthManager.startDiscovery(mActivity);
+            }
+        };
+    }
+
+    /**
+     * Checks to see if PIN for smartcard is correct.
+     * If so, proceed to attempt authentication.
+     * Otherwise, handle the incorrect PIN based on how many PIN attempts are remaining.
+     * @param pin char array containing PIN attempt.
+     * @param certDetails ICertDetails of the selected certificate from the SmartcardCertPickerDialog.
+     * @param request ClientCertRequest received from AzureActiveDirectoryWebViewClient.onReceivedClientCertRequest.
+     * @param session An ISmartcardSession created to help with interactions pertaining to certificates.
+     */
+    @RequiresApi(api = Build.VERSION_CODES.LOLLIPOP)
+    @Override
+    protected void tryUsingSmartcardWithPin(@NonNull final char[] pin,
+                                            @NonNull final ICertDetails certDetails,
+                                            @NonNull final ClientCertRequest request,
+                                            @NonNull final ISmartcardSession session)
+            throws Exception {
+        final String methodTag = TAG + ":tryUsingSmartcardWithPin";
+        if (session.verifyPin(pin)) {
+            //If pin is successfully verified, we will use the certificate to perform the rest of the logic for authentication.
+            useSmartcardCertForAuth(certDetails, pin, session, request);
+            return;
+        }
+        final int attemptsRemaining = session.getPinAttemptsRemaining();
+        mSmartcardCertBasedAuthManager.stopDiscovery(mActivity);
```

Comment:
move the implementation of this method to the abstract class, and invoke onGetCertSessionExtendedLogic in the  attemptsRemaining==0 block instead?

Reply:
what would happen if stopDiscovery() is invoked in the same pattern as the usb one? (That way we don't need to add the logic, and the code would be more sharable between nfc/usb.